File: preprocess_c/test/test_signal_state.c
/**
 * test_signal_state.c — Unit tests for SignalState.
 *
 * Compile & run:
 *   make test_signal_state && ./test_signal_state
 */
#include "../signal_state.h"
#include "../spn_spec.h"

#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int main(void)
{
    SignalState s;
    float incoming[NUM_SIGNALS];
    float row[NUM_SIGNALS];

    /* --- init: all NAN, not ready --- */
    signal_state_init(&s);
    assert(!signal_state_ready(&s));
    for (uint8_t i = 0; i < NUM_SIGNALS; i++)
        assert(isnan(s.values[i]));
    printf("PASS  init: all NAN, not ready\n");

    /* --- update with all-NAN incoming changes nothing --- */
    for (uint8_t i = 0; i < NUM_SIGNALS; i++) incoming[i] = NAN;
    signal_state_update(&s, incoming);
    assert(!signal_state_ready(&s));
    printf("PASS  update all-NAN → still not ready\n");

    /* --- update sets only the non-NAN slots --- */
    for (uint8_t i = 0; i < NUM_SIGNALS; i++) incoming[i] = NAN;
    incoming[0] = 800.0f;   /* engine_speed */
    incoming[5] = 60.0f;    /* wheel_speed */
    signal_state_update(&s, incoming);
    assert(!isnan(s.values[0]));
    assert(s.values[0] == 800.0f);
    assert(!isnan(s.values[5]));
    assert(s.values[5] == 60.0f);
    assert(isnan(s.values[1]));  /* still unseen */
    assert(!signal_state_ready(&s));
    printf("PASS  partial update sets correct slots\n");

    /* --- fill all signals → ready --- */
    for (uint8_t i = 0; i < NUM_SIGNALS; i++) incoming[i] = (float)i;
    signal_state_update(&s, incoming);
    assert(signal_state_ready(&s));
    printf("PASS  all slots filled → ready\n");

    /* --- row copies all values --- */
    signal_state_row(&s, row);
    for (uint8_t i = 0; i < NUM_SIGNALS; i++)
        assert(row[i] == s.values[i]);
    printf("PASS  signal_state_row copies all values\n");

    /* --- update overwrites only non-NAN slots, leaves others intact --- */
    for (uint8_t i = 0; i < NUM_SIGNALS; i++) incoming[i] = NAN;
    incoming[3] = 42.5f;  /* accel_pedal */
    signal_state_update(&s, incoming);
    assert(s.values[3] == 42.5f);
    assert(s.values[0] == 0.0f);   /* set from previous update, unchanged */
    printf("PASS  update with partial incoming preserves other slots\n");

    /* --- init resets to NAN / not-ready --- */
    signal_state_init(&s);
    assert(!signal_state_ready(&s));
    for (uint8_t i = 0; i < NUM_SIGNALS; i++)
        assert(isnan(s.values[i]));
    printf("PASS  re-init resets all to NAN\n");

    printf("\nAll signal_state tests passed.\n");
    return 0;
}

File: tests/test_signal_state.py
from preprocess.features.signal_state import SIGNALS, SignalState


def test_holds_latest_value_per_signal():
    s = SignalState()
    s.update({"engine_speed": 1000.0})
    s.update({"engine_speed": 1200.0})
    assert s.row()[SIGNALS.index("engine_speed")] == 1200.0


def test_unknown_signal_is_ignored():
    s = SignalState()
    s.update({"not_a_signal": 5.0})
    assert all(v is None for v in s.row())


def test_ready_only_when_all_signals_seen():
    s = SignalState()
    assert not s.ready()
    for name in SIGNALS:
        s.update({name: 1.0})
    assert s.ready()
    assert s.row() == [1.0] * len(SIGNALS)


def test_signals_come_from_spec():
    assert "engine_speed" in SIGNALS
    assert "wheel_speed" in SIGNALS
    assert "yaw_rate" in SIGNALS
    assert len(SIGNALS) == 17


def test_values_gives_names_and_skips_what_has_not_arrived():
    s = SignalState()
    s.update({"engine_speed": 1000.0, "wheel_speed": 50.0})
    assert s.values() == {"engine_speed": 1000.0, "wheel_speed": 50.0}
